src/Language/Haskell/LSP/Core.hs

```diff
@@ -831,67 +840,94 @@ initializeRequestHandler' onStartup mHandler tvarCtx req@(J.RequestMessage _ ori
-              , J._experimental                     = Nothing :: Maybe J.Value
-              }
-
-          -- TODO: wrap this up into a fn to create a response message
-          res  = J.ResponseMessage "2.0" (J.responseId origId) (Just $ J.InitializeResponseCapabilities capa) Nothing
+-- | Infers the capabilities based on registered handlers, and sets the appropriate options.
+serverCapabilities :: Options -> Handlers -> J.InitializeResponseCapabilitiesInner
+serverCapabilities o h =
+  J.InitializeResponseCapabilitiesInner
+    { J._textDocumentSync                 = sync
+    , J._hoverProvider                    = supported (hoverHandler h)
+    , J._completionProvider               = completionProvider
+    , J._signatureHelpProvider            = signatureHelpProvider
+    , J._definitionProvider               = supported (definitionHandler h)
+    , J._typeDefinitionProvider           = Just $ J.GotoOptionsStatic $ isJust $ typeDefinitionHandler h
+    , J._implementationProvider           = Just $ J.GotoOptionsStatic $ isJust $ typeDefinitionHandler h
+    , J._referencesProvider               = supported (referencesHandler h)
+    , J._documentHighlightProvider        = supported (documentHighlightHandler h)
+    , J._documentSymbolProvider           = supported (documentSymbolHandler h)
+    , J._workspaceSymbolProvider          = supported (workspaceSymbolHandler h)
+    , J._codeActionProvider               = codeActionProvider
+    , J._codeLensProvider                 = Just $ J.CodeLensOptions $ supported $ codeLensResolveHandler h
+    , J._documentFormattingProvider       = supported (documentFormattingHandler h)
+    , J._documentRangeFormattingProvider  = supported (documentRangeFormattingHandler h)
+    , J._documentOnTypeFormattingProvider = documentOnTypeFormattingProvider
+    , J._renameProvider                   = Just $ J.RenameOptionsStatic $ isJust $ renameHandler h
+    , J._documentLinkProvider             = Just $ J.DocumentLinkOptions $ Just $ isJust $ documentLinkResolveHandler h
+    , J._colorProvider                    = Just $ J.ColorOptionsStatic $ isJust $ documentColorHandler h
+    , J._foldingRangeProvider             = Just $ J.FoldingRangeOptionsStatic $ isJust $ foldingRangeHandler h
+    , J._executeCommandProvider           = executeCommandProvider
+    , J._workspace                        = Just workspace
+    -- TODO: Add something for experimental
+    , J._experimental                     = Nothing :: Maybe J.Value
+    }
+  where
+    supported (Just _) = Just True
+    supported Nothing   = Nothing
+
+    singleton :: a -> [a]
+    singleton x = [x]
+
+    completionProvider
+      | isJust $ completionHandler h = Just $
+          J.CompletionOptions
+            (Just $ isJust $ completionResolveHandler h)
+            (map singleton <$> completionTriggerCharacters o)
+            (map singleton <$> completionAllCommitCharacters o)
+      | otherwise = Nothing
+
+    codeActionProvider
```

Comment:
This breaks `ghcide` in a way that I don’t know how to fix. We currently set this to `CodeActionOptionsStatic True` and as mentioned above, I can’t set the individual flags properly since I learn the capabilities after I need to specify the options. Maybe for now, we could use something like
```
isJust $ codeActionHandler h = maybe (J.CodeActionOptionsStatic True) (CodeActionOptions . Just) (codeActionKinds o)
```

Reply:
That sounds reasonable. I left it this way as it is possible to return a `CodeActionOptions` with a null `CodeActionKinds` inside it (which doesn't make much sense given the previous comment), so I'm happy enough to default to the static capability if the kinds in the LSP options are set to Nothing